File: TOMIMSV4/DAL/Transacciones/Stock/clsLnStock_se.vb
Imports System.Data.SqlClient

Public Class clsLnStock_se

    Public Shared Sub Cargar(ByRef oBeStock_se As clsBeStock_se, ByRef dr As DataRow)
        Try
            With oBeStock_se
                .IdStockSe = IIf(IsDBNull(dr.Item("IdStockSe")), 0, dr.Item("IdStockSe"))
                .IdStock = IIf(IsDBNull(dr.Item("IdStock")), 0, dr.Item("IdStock"))
                .IdProductoBodega = IIf(IsDBNull(dr.Item("IdProductoBodega")), 0, dr.Item("IdProductoBodega"))
                .NoSerie = IIf(IsDBNull(dr.Item("NoSerie")), "", dr.Item("NoSerie"))
                .NoSerieInicial = IIf(IsDBNull(dr.Item("NoSerieInicial")), "", dr.Item("NoSerieInicial"))
                .NoSerieFinal = IIf(IsDBNull(dr.Item("NoSerieFinal")), "", dr.Item("NoSerieFinal"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
            End With
        Catch ex As Exception
            Throw ex
        End Try

    End Sub

    Public Shared Function Insertar(ByRef oBeStock_se As clsBeStock_se, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand

        Try

            Ins.Init("stock_se")
            Ins.Add("idstockse", "@idstockse", DataType.Parametro)
            Ins.Add("idstock", "@idstock", DataType.Parametro)
            Ins.Add("idproductobodega", "@idproductobodega", DataType.Parametro)
            Ins.Add("noserie", "@noserie", DataType.Parametro)
            Ins.Add("noserieinicial", "@noserieinicial", DataType.Parametro)
            Ins.Add("noseriefinal", "@noseriefinal", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)

            Dim sp As String = Ins.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe))
            cmd.Parameters.Add(New SqlParameter("@IDSTOCK", oBeStock_se.IdStock))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOBODEGA", oBeStock_se.IdProductoBodega))
            cmd.Parameters.Add(New SqlParameter("@NOSERIE", IIf(oBeStock_se.NoSerie = String.Empty, DBNull.Value, oBeStock_se.NoSerie)))
            cmd.Parameters.Add(New SqlParameter("@NOSERIEINICIAL", IIf(oBeStock_se.NoSerieInicial = String.Empty, DBNull.Value, oBeStock_se.NoSerieInicial)))
            cmd.Parameters.Add(New SqlParameter("@NOSERIEFINAL", IIf(oBeStock_se.NoSerieFinal = String.Empty, DBNull.Value, oBeStock_se.NoSerieFinal)))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeStock_se.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeStock_se.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeStock_se.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeStock_se.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeStock_se.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeStock_se As clsBeStock_se, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand

        Try

            Upd.Init("stock_se")
            Upd.Add("idstockse", "@idstockse", DataType.Parametro)
            Upd.Add("idstock", "@idstock", DataType.Parametro)
            Upd.Add("idproductobodega", "@idproductobodega", DataType.Parametro)
            Upd.Add("noserie", "@noserie", DataType.Parametro)
            Upd.Add("noserieinicial", "@noserieinicial", DataType.Parametro)
            Upd.Add("noseriefinal", "@noseriefinal", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Where("IdStockSe = @IdStockSe")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe))
            cmd.Parameters.Add(New SqlParameter("@IDSTOCK", oBeStock_se.IdStock))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOBODEGA", oBeStock_se.IdProductoBodega))
            cmd.Parameters.Add(New SqlParameter("@NOSERIE", IIf(oBeStock_se.NoSerie = String.Empty, DBNull.Value, oBeStock_se.NoSerie)))
            cmd.Parameters.Add(New SqlParameter("@NOSERIEINICIAL", IIf(oBeStock_se.NoSerieInicial = String.Empty, DBNull.Value, oBeStock_se.NoSerieInicial)))
            cmd.Parameters.Add(New SqlParameter("@NOSERIEFINAL", IIf(oBeStock_se.NoSerieFinal = String.Empty, DBNull.Value, oBeStock_se.NoSerieFinal)))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeStock_se.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeStock_se.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeStock_se.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeStock_se.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeStock_se.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Eliminar(ByRef oBeStock_se As clsBeStock_se, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            cmd.CommandType = CommandType.Text

            Dim sp As String = " Delete from Stock_se" &
             "  Where(IdStockSe = @IdStockSe)"


            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Eliminar_By_IdStock(ByVal pIdStock As Integer, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            cmd.CommandType = CommandType.Text

            Dim sp As String = " Delete from Stock_se Where(IdStock = @IdStock)"

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IdStock", pIdStock))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Obtener(ByRef oBeStock_se As clsBeStock_se) As Boolean

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Obtener = False

        Try

            lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)

            Dim sp As String = "SELECT * FROM Stock_se" &
            " Where(IdStockSe = @IdStockSe)"

            Dim cmd As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe))

            Dim dt As New DataTable
            dad.Fill(dt)

            cmd.Dispose()
            dad.Dispose()

            If dt.Rows.Count = 1 Then
                Cargar(oBeStock_se, dt.Rows(0))
                Obtener = True
            End If

            lTransaction.Commit()

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
        End Try

    End Function

End Class

File: WMS.DALCore/Stock/clsLnStock_se.cs
using AppGlobal;
using Microsoft.Data.SqlClient;
using System.Data;

namespace WMS.EntityCore.Stock
{
    public static class clsLnStock_se
    {
        private static clsInsert Ins = new clsInsert();
        private static clsUpdate Upd = new clsUpdate();
        public static void Cargar(ref clsBeStock_se oBeStock_se, DataRow dr)
        {
            if (oBeStock_se == null)
                throw new ArgumentNullException(nameof(oBeStock_se));

            if (dr == null)
                throw new ArgumentNullException(nameof(dr));

            try
            {
                // Función local para manejar valores de forma segura
                T GetSafeValue<T>(string columnName, T defaultValue, Func<object, T> converter)
                {
                    if (dr.Table.Columns.Contains(columnName) && !Convert.IsDBNull(dr[columnName]))
                    {
                        var value = dr[columnName];
                        if (value != null)
                            return converter(value);
                    }
                    return defaultValue;
                }

                oBeStock_se.IdStockSe = GetSafeValue("IdStockSe", 0, Convert.ToInt32);
                oBeStock_se.IdStock = GetSafeValue("IdStock", 0, Convert.ToInt32);
                oBeStock_se.IdProductoBodega = GetSafeValue("IdProductoBodega", 0, Convert.ToInt32);
                oBeStock_se.NoSerie = GetSafeValue("NoSerie", "", Convert.ToString) ?? "";
                oBeStock_se.NoSerieInicial = GetSafeValue("NoSerieInicial", "", Convert.ToString) ?? "";
                oBeStock_se.NoSerieFinal = GetSafeValue("NoSerieFinal", "", Convert.ToString) ?? "";
                oBeStock_se.User_agr = GetSafeValue("user_agr", "", Convert.ToString) ?? "";
                oBeStock_se.Fec_agr = GetSafeValue("fec_agr", DateTime.Now, Convert.ToDateTime);
                oBeStock_se.User_mod = GetSafeValue("user_mod", "", Convert.ToString) ?? "";
                oBeStock_se.Fec_mod = GetSafeValue("fec_mod", DateTime.Now, Convert.ToDateTime);
                oBeStock_se.Activo = GetSafeValue("activo", false, Convert.ToBoolean);
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static int Insertar(clsBeStock_se oBeStock_se, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (oBeStock_se == null)
                throw new ArgumentNullException(nameof(oBeStock_se));

            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                Ins.Init("stock_se");
                Ins.Add("idstockse", "@idstockse", "F");
                Ins.Add("idstock", "@idstock", "F");
                Ins.Add("idproductobodega", "@idproductobodega", "F");
                Ins.Add("noserie", "@noserie", "F");
                Ins.Add("noserieinicial", "@noserieinicial", "F");
                Ins.Add("noseriefinal", "@noseriefinal", "F");
                Ins.Add("user_agr", "@user_agr", "F");
                Ins.Add("fec_agr", "@fec_agr", "F");
                Ins.Add("user_mod", "@user_mod", "F");
                Ins.Add("fec_mod", "@fec_mod", "F");
                Ins.Add("activo", "@activo", "F");

                string sp = Ins.SQL();

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;

                    cmd.Parameters.Add(new SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe));
                    cmd.Parameters.Add(new SqlParameter("@IDSTOCK", oBeStock_se.IdStock));
                    cmd.Parameters.Add(new SqlParameter("@IDPRODUCTOBODEGA", oBeStock_se.IdProductoBodega));
                    cmd.Parameters.Add(new SqlParameter("@NOSERIE", string.IsNullOrEmpty(oBeStock_se.NoSerie) ? DBNull.Value : oBeStock_se.NoSerie));
                    cmd.Parameters.Add(new SqlParameter("@NOSERIEINICIAL", string.IsNullOrEmpty(oBeStock_se.NoSerieInicial) ? DBNull.Value : oBeStock_se.NoSerieInicial));
                    cmd.Parameters.Add(new SqlParameter("@NOSERIEFINAL", string.IsNullOrEmpty(oBeStock_se.NoSerieFinal) ? DBNull.Value : oBeStock_se.NoSerieFinal));
                    cmd.Parameters.Add(new SqlParameter("@USER_AGR", oBeStock_se.User_agr));
                    cmd.Parameters.Add(new SqlParameter("@FEC_AGR", oBeStock_se.Fec_agr));
                    cmd.Parameters.Add(new SqlParameter("@USER_MOD", oBeStock_se.User_mod));
                    cmd.Parameters.Add(new SqlParameter("@FEC_MOD", oBeStock_se.Fec_mod));
                    cmd.Parameters.Add(new SqlParameter("@ACTIVO", oBeStock_se.Activo));

                    return cmd.ExecuteNonQuery();
                }
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static int Actualizar(clsBeStock_se oBeStock_se, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (oBeStock_se == null)
                throw new ArgumentNullException(nameof(oBeStock_se));

            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                Upd.Init("stock_se");
                Upd.Add("idstockse", "@idstockse", "F");
                Upd.Add("idstock", "@idstock", "F");
                Upd.Add("idproductobodega", "@idproductobodega", "F");
                Upd.Add("noserie", "@noserie", "F");
                Upd.Add("noserieinicial", "@noserieinicial", "F");
                Upd.Add("noseriefinal", "@noseriefinal", "F");
                Upd.Add("user_agr", "@user_agr", "F");
                Upd.Add("fec_agr", "@fec_agr", "F");
                Upd.Add("user_mod", "@user_mod", "F");
                Upd.Add("fec_mod", "@fec_mod", "F");
                Upd.Add("activo", "@activo", "F");
                Upd.Where("IdStockSe = @IdStockSe");

                string sp = Upd.SQL();

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;

                    cmd.Parameters.Add(new SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe));
                    cmd.Parameters.Add(new SqlParameter("@IDSTOCK", oBeStock_se.IdStock));
                    cmd.Parameters.Add(new SqlParameter("@IDPRODUCTOBODEGA", oBeStock_se.IdProductoBodega));
                    cmd.Parameters.Add(new SqlParameter("@NOSERIE", string.IsNullOrEmpty(oBeStock_se.NoSerie) ? DBNull.Value : oBeStock_se.NoSerie));
                    cmd.Parameters.Add(new SqlParameter("@NOSERIEINICIAL", string.IsNullOrEmpty(oBeStock_se.NoSerieInicial) ? DBNull.Value : oBeStock_se.NoSerieInicial));
                    cmd.Parameters.Add(new SqlParameter("@NOSERIEFINAL", string.IsNullOrEmpty(oBeStock_se.NoSerieFinal) ? DBNull.Value : oBeStock_se.NoSerieFinal));
                    cmd.Parameters.Add(new SqlParameter("@USER_AGR", oBeStock_se.User_agr));
                    cmd.Parameters.Add(new SqlParameter("@FEC_AGR", oBeStock_se.Fec_agr));
                    cmd.Parameters.Add(new SqlParameter("@USER_MOD", oBeStock_se.User_mod));
                    cmd.Parameters.Add(new SqlParameter("@FEC_MOD", oBeStock_se.Fec_mod));
                    cmd.Parameters.Add(new SqlParameter("@ACTIVO", oBeStock_se.Activo));

                    return cmd.ExecuteNonQuery();
                }
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static int Eliminar(clsBeStock_se oBeStock_se, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (oBeStock_se == null)
                throw new ArgumentNullException(nameof(oBeStock_se));

            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                const string sp = @"DELETE FROM Stock_se 
                               WHERE IdStockSe = @IdStockSe";

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;
                    cmd.Parameters.Add(new SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe));
                    return cmd.ExecuteNonQuery();
                }
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static int Eliminar_By_IdStock(int pIdStock, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                const string sp = @"DELETE FROM Stock_se 
                               WHERE IdStock = @IdStock";

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;
                    cmd.Parameters.Add(new SqlParameter("@IdStock", pIdStock));
                    return cmd.ExecuteNonQuery();
                }
            }
            catch (Exception)
            {
                throw;
            }
        }

        public static bool Obtener(clsBeStock_se oBeStock_se, SqlConnection pConection, SqlTransaction pTransaction)
        {
            if (oBeStock_se == null)
                throw new ArgumentNullException(nameof(oBeStock_se));

            if (pConection == null)
                throw new ArgumentNullException(nameof(pConection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                const string sp = @"SELECT * FROM Stock_se 
                               WHERE IdStockSe = @IdStockSe";

                using (var cmd = new SqlCommand(sp, pConection, pTransaction))
                {
                    cmd.CommandType = CommandType.Text;
                    cmd.Parameters.Add(new SqlParameter("@IDSTOCKSE", oBeStock_se.IdStockSe));

                    using (var dad = new SqlDataAdapter(cmd))
                    {
                        var dt = new DataTable();
                        dad.Fill(dt);

                        if (dt.Rows.Count == 1)
                        {
                            Cargar(ref oBeStock_se, dt.Rows[0]);
                            return true;
                        }
                    }
                }

                return false;
            }
            catch (Exception)
            {
                throw;
            }
        }
        public static int Insertar_Stock_Serializado_Recepcion(clsBeStock_rec BeStockRec,
                                                               int IdStock,
                                                               SqlConnection lConnection,
                                                               SqlTransaction lTransaction)
        {
            if (BeStockRec == null)
                throw new ArgumentNullException(nameof(BeStockRec));

            if (lConnection == null)
                throw new ArgumentNullException(nameof(lConnection));

            if (lTransaction == null)
                throw new ArgumentNullException(nameof(lTransaction));

            int vRegistros = 0;

            try
            {
                int lMaxSE = MaxID(lConnection, lTransaction);
                var lStockSerializadoRec = clsLnStock_se_rec.GetAllSerieByIdStockRec(BeStockRec.IdStockRec,
                                                                                     lConnection,
                                                                                     lTransaction);

                if (lStockSerializadoRec != null && lStockSerializadoRec.Count > 0)
                {
                    foreach (var Stock_Se_Rec in lStockSerializadoRec)
                    {
                        if (Stock_Se_Rec == null)
                            continue;

                        var ObjS = new clsBeStock_se();
                        clsPublic.CopyObject(Stock_Se_Rec, ref ObjS);

                        lMaxSE += 1;

                        ObjS.IdStock = IdStock;
                        ObjS.IdStockSe = lMaxSE;

                        Stock_Se_Rec.Regularizado = true;
                        Stock_Se_Rec.Fecha_regularizacion = DateTime.Now;

                        vRegistros += Insertar(ObjS, lConnection, lTransaction);
                    }
                }

                return vRegistros;
            }
            catch (Exception)
            {                
                throw;
            }
        }

        public static int MaxID(SqlConnection pConnection, SqlTransaction pTransaction)
        {
            if (pConnection == null)
                throw new ArgumentNullException(nameof(pConnection));

            if (pTransaction == null)
                throw new ArgumentNullException(nameof(pTransaction));

            try
            {
                const string vSQL = "SELECT ISNULL(MAX(IdStockse), 0) FROM stock_se";

                using (var lCommand = new SqlCommand(vSQL, pConnection, pTransaction))
                {
                    lCommand.CommandType = CommandType.Text;
                    object lReturnValue = lCommand.ExecuteScalar();

                    if (lReturnValue != DBNull.Value && lReturnValue != null)
                    {
                        return Convert.ToInt32(lReturnValue);
                    }
                    return 0;
                }
            }
            catch (Exception)
            {
                throw;
            }
        }
    }
}